### ISSUE TITLE
CosFace ансамбль с гэпом

### ISSUE THREAD

**Issue by vshokorov** [vshokorov/power_laws_deep_ensembles#5]

**Comment by vshokorov** [comment]
![image](https://user-images.githubusercontent.com/48180167/195099527-2c10e450-dff2-419e-86b2-e6e27ae7e859.png)

**Comment by vshokorov** [comment]
# Графики ансамблей на CIFAR10:
Обучалось 5 ансамблей каждого типа, в каждом ансамбле 15 моделей. Гэп используется как дополнительная поправка к основному гэпу. 

* **simple_gap** : марджин прошлой сети это гэп следующей
* **simple_gap_marginscale_2** : то же что и simple_gap, но гэп увеличен в 2 раза
* **None** : без гэпа

## обучение без аугментации:

Общие графики:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/195991573-88314f00-37a2-4730-bb76-6b8827b22322.png">

</details>
</p>

Графики для последнего ансамбля:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/195991616-c04a06a8-7629-4604-947a-fb114e9a832b.png">
<img src="https://user-images.githubusercontent.com/48180167/195991628-b2d599a2-86d9-4ddd-a724-d0813b8b6a0c.png">
<img src="https://user-images.githubusercontent.com/48180167/195991674-f00657b5-f67a-42ca-a40c-7cd2b45c4170.png">
<img src="https://user-images.githubusercontent.com/48180167/195991683-e4c11856-fb78-4292-8476-832cf25d32ca.png">
<img src="https://user-images.githubusercontent.com/48180167/195991698-4de1b8e9-cbae-4182-ac8c-028bedf263e9.png">
<img src="https://user-images.githubusercontent.com/48180167/195991714-2f8b4b6c-1286-41ad-b7e8-ac1c4a30e832.png">

</details>
</p>

**Comment by vshokorov** [comment]
# Графики ансамблей на AGEDB:
Обучалось 5 ансамблей каждого типа, в каждом ансамбле 15 моделей. Гэп используется как дополнительная поправка к основному гэпу. 

* **simple_gap** : марджин прошлой сети это гэп следующей
* **simple_gap_marginscale_2** : то же что и simple_gap, но гэп увеличен в 2 раза
* **None** : без гэпа

## обучение без аугментации:

Общие графики:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/196718168-f491efda-22d4-4bf6-9fc4-a8fce5418e20.png">
<img src="https://user-images.githubusercontent.com/48180167/196718124-a3ac84db-e427-4f98-ae5f-1419ba85ad8e.png">


</details>
</p>

Графики для последнего ансамбля:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/196720097-16ae0b73-f0b3-40f3-b23b-2cfe762b9abb.png">
<img src="https://user-images.githubusercontent.com/48180167/196719755-c3ae54fc-8e47-4a90-b7a8-5d998dc86af8.png">
<img src="https://user-images.githubusercontent.com/48180167/196719805-66e2b34b-8595-4311-a98c-3644a37f4def.png">
<img src="https://user-images.githubusercontent.com/48180167/196719842-c77be6ee-547a-4ab4-9edb-e86c53b8579c.png">
<img src="https://user-images.githubusercontent.com/48180167/196720468-8e48217d-620b-4d99-bdd5-3e6e4bb73a99.png">
<img src="https://user-images.githubusercontent.com/48180167/196720410-1c37a09b-5470-4ec4-a7d4-ea3f25894675.png">
<img src="https://user-images.githubusercontent.com/48180167/196720545-bd79eca3-95e4-4288-9006-cdf8619de1ea.png">
<img src="https://user-images.githubusercontent.com/48180167/196720587-dcccd280-b665-4ae0-b622-336adcb8a149.png">






</details>
</p>

**Comment by vshokorov** [comment]
# Нужна ли L2 нормировка перед SoftMax в CosFace? 
 - Да, потому что хоть и во время обучения модель без softmax показывает себя лучше по точности, она плохо работает с новыми классами. Скорее всего потому, что получается слишком много степеней свободы. Работая на сфере мы явно ограничиваем регион, в которой находятся эмбеддинги. В противном случае получается бесконечное пространство. Аналогия с AE vs VAE.

![image](https://user-images.githubusercontent.com/48180167/198873199-f022611d-8400-4336-a07f-56599b7e0b2a.png)

<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/198873107-1fd37930-3fac-45a4-9270-4b0351173cb6.png">
<img src="https://user-images.githubusercontent.com/48180167/198873109-bca88814-f44c-46e3-8ba1-0c0a50b3f15d.png">
<img src="https://user-images.githubusercontent.com/48180167/198873112-5bc7b295-8ecb-408c-b012-d2a00eb329c8.png">
<img src="https://user-images.githubusercontent.com/48180167/198873115-4f88a5a1-2c7d-4617-badb-6ef0317cb887.png">

</details>
</p>

**Comment by vshokorov** [comment]
Графики построены по CrossEntropy с теми же параметрами, с какими и обучались модели.

![image](https://user-images.githubusercontent.com/48180167/198886519-2828727d-82dc-47ec-b84a-b67a6c6be5d5.png)
![image](https://user-images.githubusercontent.com/48180167/198886562-b584fb86-f68e-41c2-9980-85225a66031b.png)

Пример cosine similarity на CIFAR10, CosFace:
```
logits:                                           # entropy after Softmax: 
[-0.2620, -0.2544, -0.2098,  0.9988, -0.2732],    # entropy: 0.05620341
[-0.1757,  0.9950, -0.2479, -0.2794, -0.2848],    # entropy: 0.05854966
[ 0.8833,  0.2315, -0.3968, -0.3866, -0.3272],    # entropy: 0.19758901
```

**Comment by vshokorov** [comment]
# Графики ансамблей на ImageNet (просэмплированный по 20 картинок на класс во время обучения) :
Обучалось 4 ансамблей каждого типа, в каждом ансамбле 20 моделей. Гэп используется как дополнительная поправка к основному гэпу. 

## обучение без аугментации:

Общие графики:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/201652308-88510753-bd3d-49d1-8c35-1c8cc76ba9db.png">
<img src="https://user-images.githubusercontent.com/48180167/201652319-082c15b6-3057-419a-bf4e-263de94e7338.png">

</details>
</p>

Графики для последнего ансамбля:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/201652439-3621353a-3d0d-4b40-b4dc-7a2fc46fc1a5.png">
<img src="https://user-images.githubusercontent.com/48180167/201652463-34860a47-8025-4615-8c8f-8a27d0c9289e.png">
<img src="https://user-images.githubusercontent.com/48180167/201652505-46a3774b-98e1-4f22-9563-42d86a9f55af.png">
<img src="https://user-images.githubusercontent.com/48180167/201652548-c98adf51-b405-4bc6-8411-4d6ca696dd55.png">
<img src="https://user-images.githubusercontent.com/48180167/201652633-b9adc4b2-72c1-4735-bace-d6442b4abafa.png">
<img src="https://user-images.githubusercontent.com/48180167/201652707-cf50d774-94e8-494a-a140-c752f22aa022.png">
<img src="https://user-images.githubusercontent.com/48180167/201652891-dc9c7843-8890-4ee3-91b7-c3129b71bcd4.png">
<img src="https://user-images.githubusercontent.com/48180167/201652922-cd5738c8-9867-4a23-9b41-d14c09d3c231.png">

</details>
</p>

**Comment by vshokorov** [comment]
![image](https://user-images.githubusercontent.com/48180167/201645900-a9dba117-66cf-4a1e-b702-b0bfc9c210b2.png)

предполагаю, что это просто специфика модели. Т.е. добавил 3 графика:
1) cos_other - средний косинус до всех чужих векторов
2) margin - считается в косинусах
3) cos_self - косинус до своего вектора.


Думаю тут есть несколько этапов:
1) модель активно подтягивает вектора словаря к векторам класса ~до 400 шага. Точность быстро растет
2) Затем модель начинает расталкивать классы друг между другом. Получается, что cos_other начинает падать, что ведет к тому, расталкивая классы мы чуть чуть расталкиваем еще и свои вектора. (оттолкнули чужих, а они потянули за собой чуть чуть наших). Падает точность и марджин
3) дальше сеть находит фичи различающие классы и выходит на нормальный уровень качества

**Comment by vshokorov** [comment]
# Графики ансамблей на ImageNet (просэмплированный по 20 картинок на класс во время обучения) :
в каждом ансамбле 10 моделей. Гэп используется как дополнительная поправка к основному гэпу. 

1. **nonorm_gap_None** - бейзлайн (модель обучалась стандартным софтмаксом - нет нормировка (проекция на единичную сферу), гэпа (в CosFace во время обучения добавляется константный гэп(является просто гиперпараметром) ), температура=1)
2. **gap_None_t1** - добавили к 1. нормировку
3. **gap_None_t0.05** - добавили к 2. еще и температуру=20
4. **gap_const_gap** - добавили к 3. еще и константный гэп (получили чистый CosFace).

Далее идут ансамбли с бустингом:
добавил в название **gap_mean_cummulative** к соответствующим моделям.

![image](https://user-images.githubusercontent.com/48180167/203139991-6730dbed-7a8b-490d-b058-7854a7489fba.png)
(ансамбли gap_None_t1 имеют низкую точность,  поэтому вырезал их из графика)

![image](https://user-images.githubusercontent.com/48180167/203583439-b51f943f-100d-42be-8572-8003360f9662.png)
точность на тесте первых трех моделей независимо (модели не ансамблируются).

Дополнительные графики:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/203139861-3648c1e4-822c-446e-865b-760cbe4fa366.png">
<img src="https://user-images.githubusercontent.com/48180167/203139904-425f4bd8-b457-4559-9f2a-35c9f7b317ff.png">

<img src="https://user-images.githubusercontent.com/48180167/203139925-03e28b16-7d84-4175-bbae-58731ccb155b.png">
<img src="https://user-images.githubusercontent.com/48180167/203139967-4b0da813-37e0-4d1e-9dd0-ec11127e8bfc.png">
<img src="https://user-images.githubusercontent.com/48180167/203140016-b98a47ba-7808-491a-a8e1-623471b2b46c.png">
<img src="https://user-images.githubusercontent.com/48180167/203140028-c1608966-d2e9-4bbb-b2e7-e1ad5d13b747.png">


</details>
</p>

**Comment by vshokorov** [comment]
# Графики ансамблей на ImageNet (просэмплированный по 20 картинок на класс):
Во время обучения, использовалось 800 классов, тестируется на оставшихся 200х. В каждом ансамбле 10 моделей. Гэп используется как дополнительная поправка к основному гэпу. 

Для теста используется косинусное сходство двух изображений. Если оно больше определенного порога, то эти изображения из одного класса. Число позитивных и негативных пары сбалансировано, т.е. случайное угадывание даст точность 0.5. Если использовать _скалярное произведение_ как меру сходства изображений для ансамблей с "nonorm", то точность будет 0.5.

<p>
<img src="https://user-images.githubusercontent.com/48180167/203577475-c3c3a20b-303e-4a96-a49f-791dd35c85be.png">
<img src="https://user-images.githubusercontent.com/48180167/203577498-2547a025-ecd2-4a5f-a6f6-b49ab3435d70.png">
</p>


![image](https://user-images.githubusercontent.com/48180167/203584822-239cee71-c993-4e8b-9517-86c9224505fe.png)
точность моделей на тесте независимо (модели не ансамблируются).

Дополнительные графики:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/203578298-44fdb659-9104-4fa9-ac81-b83657250e3d.png">
<img src="https://user-images.githubusercontent.com/48180167/203578314-61260afc-0e5a-4896-b815-a7eea9e5bc2d.png">

<img src="https://user-images.githubusercontent.com/48180167/203578226-bd642810-06db-4394-a626-ba77ba4d5b21.png">

</details>
</p>

**Comment by vshokorov** [comment]
# Графики ансамблей на AGEDB (просэмплированный по 20 картинок на класс):
## Классы на тесте и трейне совпадают

![image](https://user-images.githubusercontent.com/48180167/204825586-6dbb0d8c-7abd-4c31-b169-9f92ef277870.png)

Дополнительные графики:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/204825738-44d221c7-2c13-4848-934b-4cb1bbfa9554.png">
<img src="https://user-images.githubusercontent.com/48180167/204825764-1663e271-542e-40af-908e-195b7d328370.png">

<img src="https://user-images.githubusercontent.com/48180167/204826070-db86f266-0426-4697-8969-f15a080e86a3.png">
<br>
<img src="https://user-images.githubusercontent.com/48180167/204827276-5ba2721a-01b9-424a-942d-99c4dbf4c102.png">
<img src="https://user-images.githubusercontent.com/48180167/204827306-f3ecd845-1b24-4486-9f0b-f4c90c896de2.png">
<img src="https://user-images.githubusercontent.com/48180167/204827506-447321ff-ddb3-4d1e-acf6-bfd13e482537.png">
<img src="https://user-images.githubusercontent.com/48180167/204827521-00b58b34-dc3e-4ca0-98ee-4bdbcb41a533.png">


</details>
</p>

## Классы на тесте и трейне НЕ совпадают

<p>
<img src="https://user-images.githubusercontent.com/48180167/204832785-bc6ea84b-8438-45ff-beab-fcc32197c330.png">
<img src="https://user-images.githubusercontent.com/48180167/204832825-468a7cc6-b028-4dd2-87ea-2c7510dcc47c.png">
</p>


Дополнительные графики:
<p>
<details>
<summary>Click this to collapse/fold.</summary>

<img src="https://user-images.githubusercontent.com/48180167/204833014-8ed850b4-15f6-454f-8c14-ffc64740cd98.png">
<img src="https://user-images.githubusercontent.com/48180167/204833197-f62377b3-2868-47e4-8519-3fb41285e948.png">

</details>
</p>
